File: backend/src/main/java/com/innovategov/backend/service/AuthService.java
package com.innovategov.backend.service;

import com.innovategov.backend.dto.auth.AuthResponse;
import com.innovategov.backend.dto.auth.LoginRequest;
import com.innovategov.backend.dto.auth.RegisterRequest;
import com.innovategov.backend.entity.GovernmentDepartment;
import com.innovategov.backend.entity.Role;
import com.innovategov.backend.entity.Startup;
import com.innovategov.backend.entity.User;
import com.innovategov.backend.entity.enums.RoleName;
import com.innovategov.backend.exception.ApiException;
import com.innovategov.backend.repository.GovernmentDepartmentRepository;
import com.innovategov.backend.repository.RoleRepository;
import com.innovategov.backend.repository.StartupRepository;
import com.innovategov.backend.repository.UserRepository;
import com.innovategov.backend.security.JwtService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final RoleRepository roleRepository;
    private final GovernmentDepartmentRepository governmentDepartmentRepository;
    private final StartupRepository startupRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;
    private final AuditService auditService;

    @Transactional
    public AuthResponse register(RegisterRequest request) {
        if (request.getRole() == RoleName.ADMIN || request.getRole() == RoleName.EXPERT) {
            throw ApiException.forbidden("Admin and Expert accounts can only be provisioned by an administrator");
        }
        if (userRepository.existsByEmail(request.getEmail())) {
            throw ApiException.conflict("An account with this email already exists");
        }
        if (request.getRole() == RoleName.STARTUP && isBlank(request.getCompanyName())) {
            throw ApiException.badRequest("companyName is required when registering as a startup");
        }
        if (request.getRole() == RoleName.GOVERNMENT && isBlank(request.getDepartmentName())) {
            throw ApiException.badRequest("departmentName is required when registering as a government department");
        }

        Role role = roleRepository.findByName(request.getRole())
                .orElseThrow(() -> ApiException.notFound("Role not found: " + request.getRole()));

        User user = new User();
        user.setEmail(request.getEmail().toLowerCase().trim());
        user.setPasswordHash(passwordEncoder.encode(request.getPassword()));
        user.setFullName(request.getFullName());
        user.setRole(role);
        user.setActive(true);
        userRepository.save(user);

        if (request.getRole() == RoleName.STARTUP) {
            Startup startup = new Startup();
            startup.setUser(user);
            startup.setCompanyName(request.getCompanyName());
            startup.setReadinessScore(BigDecimal.valueOf(50));
            startupRepository.save(startup);
        } else {
            GovernmentDepartment department = new GovernmentDepartment();
            department.setUser(user);
            department.setDepartmentName(request.getDepartmentName());
            department.setMinistry(request.getMinistry());
            department.setRegion(request.getRegion());
            governmentDepartmentRepository.save(department);
        }

        auditService.log(user, "REGISTER", "User", user.getId(), Map.of("role", role.getName().name()));

        String token = jwtService.generateToken(user.getId(), user.getEmail(), role.getName().name());
        return new AuthResponse(token, user.getId(), user.getEmail(), user.getFullName(), role.getName().name());
    }

    public AuthResponse login(LoginRequest request) {
        User user = userRepository.findByEmail(request.getEmail().toLowerCase().trim())
                .orElseThrow(() -> new ApiException(HttpStatus.UNAUTHORIZED, "Invalid email or password"));

        if (!user.isActive()) {
            throw ApiException.forbidden("This account has been deactivated");
        }
        if (!passwordEncoder.matches(request.getPassword(), user.getPasswordHash())) {
            throw new ApiException(HttpStatus.UNAUTHORIZED, "Invalid email or password");
        }

        auditService.log(user, "LOGIN", "User", user.getId(), null);

        String token = jwtService.generateToken(user.getId(), user.getEmail(), user.getRole().getName().name());
        return new AuthResponse(token, user.getId(), user.getEmail(), user.getFullName(), user.getRole().getName().name());
    }

    private boolean isBlank(String s) {
        return s == null || s.isBlank();
    }
}

File: backend/app/services/auth_service.py
"""
Registration, login and the current-user lookup.

Port of Java's `AuthService`. Every guard, message and status code is
reproduced, because the frontend surfaces `message` verbatim:

* EXPERT/ADMIN self-registration -> 403
* duplicate email -> 409
* STARTUP without `companyName` / GOVERNMENT without `departmentName` -> 400
* unknown email or wrong password -> 401 "Invalid email or password"
* deactivated account -> 403

The unknown-email and wrong-password cases share one message on purpose: they
must be indistinguishable, or the endpoint becomes an account-enumeration
oracle.
"""
from __future__ import annotations

from decimal import Decimal

from sqlalchemy.orm import Session

from app.core.exceptions import ApiException
from app.models import GovernmentDepartment, RoleName, Startup, User
from app.repositories import UserRepository
from app.schemas.auth import AuthResponse, LoginRequest, RegisterRequest
from app.security.jwt import create_access_token
from app.security.password import hash_password, verify_password
from app.services.audit_service import Action, AuditService

#: Java sets this explicitly on a self-registered startup rather than relying
#: on the column default, so it is set explicitly here too.
DEFAULT_READINESS_SCORE = Decimal("50")

#: Roles a person may create for themselves. Experts and admins are provisioned.
SELF_SERVICE_ROLES = (RoleName.STARTUP, RoleName.GOVERNMENT)


class AuthService:
    def __init__(self, db: Session):
        self.db = db
        self.users = UserRepository(db)
        self.audit = AuditService(db)

    # ------------------------------------------------------------------

    def register(self, request: RegisterRequest) -> AuthResponse:
        if request.role not in SELF_SERVICE_ROLES:
            raise ApiException.forbidden(
                "Admin and Expert accounts can only be provisioned by an administrator")

        email = request.email.strip().lower()
        if self.users.email_exists(email):
            raise ApiException.conflict("An account with this email already exists")

        if request.role is RoleName.STARTUP and not (request.company_name or "").strip():
            raise ApiException.bad_request(
                "companyName is required when registering as a startup")
        if request.role is RoleName.GOVERNMENT and not (request.department_name or "").strip():
            raise ApiException.bad_request(
                "departmentName is required when registering as a government department")

        role = self.users.get_role(request.role)
        if role is None:
            raise ApiException.not_found(f"Role not found: {request.role.value}")

        user = self.users.add(User(
            email=email,
            password_hash=hash_password(request.password),
            full_name=request.full_name,
            role_id=role.id,
            is_active=True,
        ))

        if request.role is RoleName.STARTUP:
            self.users.add_startup(Startup(
                user_id=user.id,
                company_name=request.company_name,
                readiness_score=DEFAULT_READINESS_SCORE,
            ))
        else:
            self.users.add_department(GovernmentDepartment(
                user_id=user.id,
                department_name=request.department_name,
                ministry=request.ministry,
                region=request.region,
            ))

        self.audit.log(user, Action.REGISTER, "User", user.id,
                       {"role": role.name.value})
        self.db.commit()

        return AuthResponse(
            token=create_access_token(user.id, user.email, role.name.value),
            user_id=user.id,
            email=user.email,
            full_name=user.full_name,
            role=role.name,
        )

    # ------------------------------------------------------------------

    def login(self, request: LoginRequest) -> AuthResponse:
        user = self.users.find_by_email(request.email)

        # Verify against a real hash even when the account is unknown, so the
        # response time does not reveal whether the address exists.
        if user is None:
            hash_password(request.password)
            raise ApiException(401, "Invalid email or password")

        if not user.is_active:
            raise ApiException.forbidden("This account has been deactivated")

        if not verify_password(request.password, user.password_hash):
            raise ApiException(401, "Invalid email or password")

        self.audit.log(user, Action.LOGIN, "User", user.id)
        self.db.commit()

        role_name = user.role.name
        return AuthResponse(
            token=create_access_token(user.id, user.email, role_name.value),
            user_id=user.id,
            email=user.email,
            full_name=user.full_name,
            role=role_name,
        )
